### PR TITLE
Moesif reporter with polling.

### PR DESCRIPTION
**Pull request by SarithRavI**
## Purpose
Choreo platform user should be able to send API analytics to Moesif platform [[1]](https://www.moesif.com/) given that user has provided a one-time write only Moesif key [[2]](https://www.moesif.com/docs/api#moesif-api-reference).  

## New Features.
Here I explain how this PR differs from #86.

 In previous implementation:

-  We initialize the internal map explicitly at the beginning of publishing.
-  In the absence of the moesif key, the reporter will call the microservice and fetch only one/ relevant instance.

 In current implementation:

-  We make periodical calls/polling to the microservice from the start to init/refresh the entire  internal map (not single instance).
-  We temporarily queue events associated with organization of whom moesif keys are not currently in the internal map.
-  At the end of each poll to the microservice the events we stored temporally, will be dequeued to the inbound Event Queue.
- Additionally removed redundant initialization of the `MoesifClient` within `ParallelQueueWorker`.

Reason for the changes: Can't grantee there will be low traffic to the microservice. 

   
## Goals
Should be able to direct analytics events to Moesif platform securely, reliably by adhering to multi-tenant architecture.

## Approach
This implementation shares many similarities to already existing Reporters.
Only addition is Moesif reporter is communicating with an external microservice to handle end user details. 

## Automation tests
 - Unit tests 
   > Covered for Moesif event builder

### Status
Draft (Microservice should be in prod to proceed ).

### CODE REVIEW

File: component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/client/MoesifClient.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+                .headers(reqHeaders)
+                .build();
+
+        EventResponseModel eventRsp = new EventResponseBuilder()
+                .time(new Date(System.currentTimeMillis() + 1000))
+                .status((int) data.get(Constants.TARGET_RESPONSE_CODE))
+                .headers(rspHeaders)
+                .build();
+
+        return new EventBuilder()
```
Comment (SarithRavI) [inline component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/client/MoesifClient.java:182]: @ruks : no need eventModel. can return the new instance.
Comment : Ack

File: component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/client/MoesifClientContextHolder.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+package org.wso2.am.analytics.publisher.client;
+
+/**
+ * Holds context of the Moesif client retry mechanism.
+ */
+public class MoesifClientContextHolder {
+    public static final ThreadLocal<Integer> PUBLISH_ATTEMPTS = new ThreadLocal<Integer>() {
+        @Override
+        protected Integer initialValue() {
+            return Integer.valueOf(1);
```
Comment (SarithRavI) [inline component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/client/MoesifClientContextHolder.java:27]: @ruks : return 1;
comment : return 1; statement is caught by findbugs

File: pom.xml [hunk 3]
```diff
@@ -34,6 +34,11 @@
     <name>APIM Analytics Publisher</name>
     <dependencyManagement>
         <dependencies>
+            <dependency>
+            <groupId>com.moesif.api</groupId>
+            <artifactId>moesifapi</artifactId>
+            <version>${moesif.api.version}</version>
```
Comment (SarithRavI) [inline pom.xml:40]: @ruks: can you parameterize this version?
comment: Ack.

File: component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+ * Also ease Moesif SDK client initiation by maintaining a map(moesif key, moesif sdk client {@link MoesifAPIClient}).
+ */
+public class MoesifKeyRetriever {
+    private static final Logger log = LoggerFactory.getLogger(MoesifKeyRetriever.class);
+    private static MoesifKeyRetriever moesifKeyRetriever;
+    private ConcurrentHashMap<String, String> orgIDMoesifKeyMap;
+
+    private ConcurrentHashMap<String, MoesifAPIClient> moesifKeyClientMap;
+    private String gaAuthUsername;
+    private char[] gaAuthPwd;
```
Comment (SarithRavI) [inline component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java:54]: @VirajSalaka : we do not keep passwords as plain strings. Rather lets assign it as char array.
Comment : Changed type to char array

File: component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+        try {
+            con = (HttpsURLConnection) obj.openConnection();
+            con.setRequestMethod("GET");
+            con.setRequestProperty("Authorization", authHeaderValue);
+            con.setRequestProperty("Content-Type", MoesifMicroserviceConstants.CONTENT_TYPE);
+            con.setReadTimeout(MoesifMicroserviceConstants.REQUEST_READ_TIMEOUT);
+            int responseCode = con.getResponseCode();
+            if (responseCode == HttpsURLConnection.HTTP_OK) {
+                try (BufferedReader in = new BufferedReader(
+                        new InputStreamReader(con.getInputStream(), StandardCharsets.UTF_8));) {
```
Comment (SarithRavI) [inline component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java:170]: VirajSalaka : StandardCharsets.UTF_8 can be used I guess
Comment : Ack

File: component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+        HttpsURLConnection con = null;
+        try {
+            con = (HttpsURLConnection) obj.openConnection();
+            con.setRequestMethod("GET");
+            con.setRequestProperty("Authorization", authHeaderValue);
+            con.setRequestProperty("Content-Type", MoesifMicroserviceConstants.CONTENT_TYPE);
+            con.setReadTimeout(MoesifMicroserviceConstants.REQUEST_READ_TIMEOUT);
+            int responseCode = con.getResponseCode();
+            if (responseCode == HttpsURLConnection.HTTP_OK) {
+                try (BufferedReader in = new BufferedReader(
```
Comment (SarithRavI) [inline component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java:169]: @VirajSalaka : 

Let's do a try with resources here.
try(BufferedReader in = new BufferedReader(new InputStreamReader(con.getInputStream(), "UTF-8"))) {
}

then you won't need in.close in the finally block

Comment : Ack

File: component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+     * Fetches all the available orgID - MoesifKey pairs from the microservice.
+     *
+     * @throws IOException
+     * @throws APICallException
+     */
+    private void callListResource() throws IOException, APICallException {
+        final URL obj;
+        try {
+            obj = new URL(MoesifMicroserviceConstants.LIST_URL);
+        } catch (MalformedURLException ex) {
```
Comment (SarithRavI) [inline component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java:154]: @VirajSalaka : This error is not related to an event right. Rather it indicates that it cannot connect to moesif ms.
In that case lets bring in more meaningful error message indicating that the moesif MS 's url is invalid.

Comment: Ack

File: component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+                    String inputLine;
+                    StringBuffer response = new StringBuffer();
+
+                    while ((inputLine = in.readLine()) != null) {
+                        response.append(inputLine);
+                    }
+                    updateMap(response.toString());
+                }
+            } else if (responseCode >= 400 && responseCode < 500) {
+                log.error("Getting {} from the microservice.", responseCode);
```
Comment (SarithRavI) [inline component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java:180]: @VirajSalaka : Error message needs to be updated.

Comment : Ack

File: component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/exception/APICallException.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright (c) 2023, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```
Comment (SarithRavI) [inline component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/exception/APICallException.java:2]: @RakhithaRR : Incorrect year in the licence
Comment : Ack.

File: component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+
+    /**
+     * Retrieves moesif key of the given orgID by calling the microservice.
+     *
+     * @param orgID
+     * @return
+     * @throws IOException
+     * @throws APICallException
+     */
+    private String callDetailResource(String orgID) throws IOException, APICallException {
```
Comment (SarithRavI) [inline component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java:199]: @RakhithaRR : Shall we add a proper doc comment for this method?

comment : Ack

File: component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+        String moesifKey = orgIDMoesifKeyMap.remove(orgID);
+        moesifKeyClientMap.remove(moesifKey);
+    }
+
+    /**
+     * Fetches all the available orgID - MoesifKey pairs from the microservice.
+     *
+     * @throws IOException
+     * @throws APICallException
+     */
```
Comment (SarithRavI) [inline component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java:149]: @RakhithaRR : Shall we add a proper doc comment for this method?

comment : Ack.

File: component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+        HttpsURLConnection con = null;
+        try {
+            con = (HttpsURLConnection) obj.openConnection();
+            con.setRequestMethod("GET");
+            con.setRequestProperty("Content-Type", MoesifMicroserviceConstants.CONTENT_TYPE);
+            con.setRequestProperty("Authorization", authHeaderValue);
+            con.setReadTimeout(MoesifMicroserviceConstants.REQUEST_READ_TIMEOUT);
+            int responseCode = con.getResponseCode();
+            if (responseCode == HttpsURLConnection.HTTP_OK) {
+                try (BufferedReader in = new BufferedReader(
```
Comment (SarithRavI) [inline component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java:220]: @RakhithaRR : Shall we use try-with-resources approach here?
comment : Ack.

File: component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+     * Will initialize the empty orgID-MoesifKey map.
+     * Will refresh/refill the  orgID-MoesifKey map.
+     */
+    public void initOrRefreshOrgIDMoesifKeyMap() {
+        int attempts = MoesifMicroserviceConstants.NUM_RETRY_ATTEMPTS;
+        try {
+            callListResource();
+        } catch (IOException | APICallException ex) {
+            // TODO: Separate retry logic to a separate class.
+            log.error("First attempt of refreshing internal map failed,retrying.", ex);
```
Comment (SarithRavI) [inline component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java:81]: @AnuGayan : Shall we add more context in to the error message

Comment : Ack

File: component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java [hunk 14]
```diff
@@ -0,0 +1,305 @@
...
+        orgIDMoesifKeyMap.put(orgID, moesifKey);
+        moesifKeyClientMap.put(moesifKey, new MoesifAPIClient(moesifKey));
+    }
+
+    private void updateMap(String response) {
+        Gson gson = new Gson();
+        String json = response;
+
+        Type collectionType = new TypeToken<Collection<MoesifKeyEntry>>() {
+        }.getType();
```
Comment (ruks) [inline component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java:276]:
```
Type collectionType = new CustomType().getType();
```

and define new static inner class 
```
    static class CustomType extends TypeToken<Collection<MoesifKeyEntry>> {
        // nothing to add
    }
    ```

Comment (SarithRavI) [reply component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java:276]: Decided to resolve this bug by excluding associated findbug.

File: component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java (hunk 8 above)
Comment (VirajSalaka) [reply component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java:182]: lets be specific about which microservice it is.

Comment (SarithRavI) [reply component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java:182]: Ack.

File: component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java [hunk 15]
```diff
@@ -0,0 +1,305 @@
...
+        final URL obj;
+        try {
+            obj = new URL(url);
+        } catch (MalformedURLException ex) {
+            log.error("Failed calling Moesif microservice. Attempted to call url: {}", url.replaceAll("[\r\n]", ""),
+                    ex.getMessage().replaceAll("[\r\n]", ""));
+            return null;
+        }
+        String authHeaderValue = getAuthHeader(gaAuthUsername, gaAuthPwd);
+        HttpsURLConnection con = null;
```
Comment (VirajSalaka) [inline component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java:221]: Even this https connection logic also common right

Comment (SarithRavI) [reply component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java:221]: Yes Viraj
Shall I move it ?

File: component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java [hunk 16]
```diff
@@ -0,0 +1,305 @@
...
+
+                    while ((inputLine = in.readLine()) != null) {
+                        response.append(inputLine);
+                    }
+                    updateMap(response.toString());
+                }
+            } else if (responseCode >= 400 && responseCode < 500) {
+                log.error("Getting {} from the microservice.", responseCode);
+            } else {
+                throw new APICallException("Getting " + responseCode + " from the microservice and retrying.");
```
[the remark this answers is not in the document]
Comment (SarithRavI) [reply component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java:184]: Ack.

File: component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java [hunk 17]
```diff
@@ -0,0 +1,305 @@
...
+                attempts--;
+                try {
+                    Thread.sleep(MoesifMicroserviceConstants.TIME_TO_WAIT);
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+                try {
+                    callListResource();
+                } catch (IOException | APICallException e) {
+                    log.error("Retry attempt failed.", e);
```
Comment (RakhithaRR) [inline component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java:94]: Let's improve this log also with the retry attempt number and what it is we are trying to refresh here.

**Comment by SarithRavI** [comment]
There are two major updates as follows:
1. We no longer keep an internal map of Moesif key - Moesif Client
What Moesif client does is changing the Moesif key on which the APIController works on. Given that APIController is a singleton, if we were to keep an internal map of Moesif clients, APIController will work on the moesif key passed via the most recent initiation of the moesif client (in other words most recent moesif client instance inserted to the map). This not desirable.

2. Added scheduled polling mechanism 
Instead of calling the microservice each time we fail to get moesif key of a given org, we call the microservice for every 5 minutes and refresh the internal  org ID - Moesif Key map.

**Review by ruks — APPROVED** [review]
